vioscsi/helper.c

```diff
@@ -82,53 +86,55 @@ ENTER_FN_SRB();
             }
         } else {
             RhelDbgPrint(TRACE_LEVEL_ERROR, " StorPortGetStartIoPerfParams failed srb 0x%p status 0x%x MessageNumber %d.\n", Srb, status, param.MessageNumber);
+            // FIXME
+            // Should we return on this error..?
+            // Should it be TRACE_LEVEL_FATAL..?
+            //is_done = TRUE;
         }
     }
-
+    
     srbExt = SRB_EXTENSION(Srb);
-
+    
     if (!srbExt) {
         RhelDbgPrint(TRACE_LEVEL_INFORMATION, " No SRB Extenstion for SRB 0x%p \n", Srb);
+EXIT_FN_SRB();
         return;
     }
-
+    
     MessageID = QUEUE_TO_MESSAGE(QueueNumber);
     index = QueueNumber - VIRTIO_SCSI_REQUEST_QUEUE_0;
-
+    
     if (adaptExt->reset_in_progress) {
-        RhelDbgPrint(TRACE_LEVEL_FATAL, " Reset is in progress, completing SRB 0x%p with SRB_STATUS_BUS_RESET.\n", Srb);
+        RhelDbgPrint(TRACE_LEVEL_WARNING, " Reset is in progress, completing SRB 0x%p with SRB_STATUS_BUS_RESET.\n", Srb);
         SRB_SET_SRB_STATUS(Srb, SRB_STATUS_BUS_RESET);
         CompleteRequest(DeviceExtension, Srb);
+EXIT_FN_SRB();
         return;
     }
-
-    VioScsiVQLock(DeviceExtension, MessageID, &LockHandle, FALSE);
+    VioScsiLockManager(DeviceExtension, MessageID, &LockHandle, FALSE, VIOSCSI_VQLOCKOP_LOCK);
     SET_VA_PA();
     res = virtqueue_add_buf(adaptExt->vq[QueueNumber],
         srbExt->psgl,
         srbExt->out, srbExt->in,
         &srbExt->cmd, va, pa);
-
+    
     if (res >= 0) {
+        notify = virtqueue_kick_prepare(adaptExt->vq[QueueNumber]);
         element = &adaptExt->processing_srbs[index];
         InsertTailList(&element->srb_list, &srbExt->list_entry);
         element->srb_cnt++;
-    }
-    VioScsiVQUnlock(DeviceExtension, MessageID, &LockHandle, FALSE);
-    if ( res >= 0){
-        if (virtqueue_kick_prepare(adaptExt->vq[QueueNumber])) {
-            virtqueue_notify(adaptExt->vq[QueueNumber]);
-        }
     } else {
-        virtqueue_notify(adaptExt->vq[QueueNumber]);
         ScsiStatus = SCSISTAT_QUEUE_FULL;
         SRB_SET_SRB_STATUS(Srb, SRB_STATUS_BUSY);
         SRB_SET_SCSI_STATUS(Srb, ScsiStatus);
         StorPortBusy(DeviceExtension, 10);
         CompleteRequest(DeviceExtension, Srb);
-        RhelDbgPrint(TRACE_LEVEL_FATAL, " Could not put an SRB into a VQ, so complete it with SRB_STATUS_BUSY. QueueNumber = %d, SRB = 0x%p, Lun = %d, TimeOut = %d.\n", QueueNumber, srbExt->Srb, SRB_LUN(Srb), Srb->TimeOutValue);
+        RhelDbgPrint(TRACE_LEVEL_WARNING, " Could not put an SRB into a VQ, so complete it with SRB_STATUS_BUSY. QueueNumber = %d, SRB = 0x%p, Lun = %d, TimeOut = %d.\n", QueueNumber, srbExt->Srb, SRB_LUN(Srb), Srb->TimeOutValue);
+    }
+    VioScsiLockManager(DeviceExtension, MessageID, &LockHandle, FALSE, VIOSCSI_VQLOCKOP_UNLOCK);
+    if (notify){
```

Comment:
Yeah, I think there was a `notify = TRUE;` at the foot of the QUEUE_FULL branch previously. I didn't notice any change with or without it in testing. I can have another go around though, but it could be in virtqueue implementation (which I'm presently looking at).

Reply:
So it appears that res will only ever (famous last words) be a 0 (SUCCESS) or a 28 (ENOSPC: No space left on device), i.e. SCSISTAT_QUEUE_FULL. The test should really be `if (res **==** 0) {}`...

As no data was added to a buffer, does it really need any further action?

You might observe in my reply to @frwbr above that a call to `virtqueue_kick()` just calls `virtqueue_kick_prepare()` and so a `virtqueue_notify(vq)` won't happen without the former returning 0.

We could call a `virtqueue_notify(vq)`, but the device won't find anything there.

My thoughts were that it would just consume compute resources and time.